## 1. Import libraries

In [1]:
import pandas as pd
import numpy as np
import os

## 2. Flat file digestion
STRUCTURED DATASET FROM CDC WONDER

In [2]:
# load csv file
df = pd.read_csv('../data/raw/YEAR_STATE_AGE_MCD/YEAR_STATE_AGE_MCD_1999.csv')

## 3. Inspect and validate

In [3]:
# check data
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 7113 entries, 0 to 7112
Data columns (total 12 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   Notes                         75 non-null     str    
 1   Year                          7038 non-null   float64
 2   Year Code                     7038 non-null   float64
 3   State                         7038 non-null   str    
 4   State Code                    7038 non-null   float64
 5   Five-Year Age Groups          7038 non-null   str    
 6   Five-Year Age Groups Code     7038 non-null   str    
 7   Multiple Cause of death       7038 non-null   str    
 8   Multiple Cause of death Code  7038 non-null   str    
 9   Deaths                        7038 non-null   str    
 10  Population                    7038 non-null   str    
 11  Crude Rate                    7038 non-null   str    
dtypes: float64(3), str(9)
memory usage: 1.1 MB


In [4]:
# check first 5 rows
df.head()

,Notes,Year,Year Code,State,State Code,Five-Year Age Groups,Five-Year Age Groups Code,Multiple Cause of death,Multiple Cause of death Code,Deaths,Population,Crude Rate
0,NaN,1999.0,1999.0,Alabama,1.0,< 1 year,1,Opium,T40.0,0,58787,Unreliable
1,NaN,1999.0,1999.0,Alabama,1.0,< 1 year,1,Heroin,T40.1,0,58787,Unreliable
2,NaN,1999.0,1999.0,Alabama,1.0,< 1 year,1,Other opioids,T40.2,0,58787,Unreliable
3,NaN,1999.0,1999.0,Alabama,1.0,< 1 year,1,Methadone,T40.3,0,58787,Unreliable
4,NaN,1999.0,1999.0,Alabama,1.0,< 1 year,1,Other synthetic narcotics,T40.4,0,58787,Unreliable


In [5]:
# check last 5 rows
df.tail()

,Notes,Year,Year Code,State,State Code,Five-Year Age Groups,Five-Year Age Groups Code,Multiple Cause of death,Multiple Cause of death Code,Deaths,Population,Crude Rate
7108,2019 data. Underlying and multiple cause of de...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
7109,"unspecified firearm discharge, ICD-10 code X95...",NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
7110,2020.,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
7111,9. Changes to cause of death classification af...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
7112,http://wonder.cdc.gov/wonder/help/mcd.html#ICD...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


### Check missing values

In [6]:
df.isna().sum()

Notes                           7038
Year                              75
Year Code                         75
State                             75
State Code                        75
Five-Year Age Groups              75
Five-Year Age Groups Code         75
Multiple Cause of death           75
Multiple Cause of death Code      75
Deaths                            75
Population                        75
Crude Rate                        75
dtype: int64

### Check duplicate rows

In [7]:
df.duplicated().sum()

np.int64(6)

In [8]:
df[df.duplicated(keep = False)]

,Notes,Year,Year Code,State,State Code,Five-Year Age Groups,Five-Year Age Groups Code,Multiple Cause of death,Multiple Cause of death Code,Deaths,Population,Crude Rate
7038,---,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
7051,---,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
7053,---,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
7055,---,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
7060,---,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
7064,---,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
7068,---,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


### Check count of unique values across columns

In [9]:
# return no. of unique values in 'Notes' column
df.nunique()

Notes                            69
Year                              1
Year Code                         1
State                            51
State Code                       51
Five-Year Age Groups             23
Five-Year Age Groups Code        23
Multiple Cause of death           6
Multiple Cause of death Code      6
Deaths                           62
Population                      919
Crude Rate                       73
dtype: int64

### Check unique values in `Notes`

In [10]:
# return no. of unique values in 'Notes' column
df['Notes'].nunique()

69

In [11]:
# return list of unique values in 'Notes' column
list(df['Notes'].unique())

[nan,
 '---',
 'Dataset: Multiple Cause of Death, 1999-2020',
 'Query Parameters:',
 'Title: MCD_YEAR_STATE_AGE_1999',
 'MCD - ICD-10 Codes: T40.0 (Opium); T40.1 (Heroin); T40.2 (Other opioids); T40.3 (Methadone); T40.4 (Other synthetic narcotics);',
 'T40.6 (Other and unspecified narcotics)',
 'Year/Month: 1999',
 'Group By: Year; State; Five-Year Age Groups; Multiple Cause of death',
 'Show Totals: Disabled',
 'Show Zero Values: True',
 'Show Suppressed: True',
 'Calculate Rates Per: 100,000',
 'Rate Options: Default intercensal populations for years 2001-2009 (except Infant Age Groups)',
 'Help: See http://wonder.cdc.gov/wonder/help/mcd.html for more information.',
 'Query Date: May 31, 2026 12:14:44 AM',
 'Suggested Citation: Centers for Disease Control and Prevention, National Center for Health Statistics. National Vital Statistics',
 'System, Mortality 1999-2020 on CDC WONDER Online Database, released in 2021. Data are from the Multiple Cause of Death Files,',
 '1999-2020, as com

In [12]:
# return data where 'Notes' column has text
df[df['Notes'].notna()]

,Notes,Year,Year Code,State,State Code,Five-Year Age Groups,Five-Year Age Groups Code,Multiple Cause of death,Multiple Cause of death Code,Deaths,Population,Crude Rate
7038,---,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
7039,"Dataset: Multiple Cause of Death, 1999-2020",NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
7040,Query Parameters:,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
7041,Title: MCD_YEAR_STATE_AGE_1999,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
7042,MCD - ICD-10 Codes: T40.0 (Opium); T40.1 (Hero...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...
7108,2019 data. Underlying and multiple cause of de...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
7109,"unspecified firearm discharge, ICD-10 code X95...",NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
7110,2020.,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
7111,9. Changes to cause of death classification af...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [13]:
# return indexes of rows where 'Notes' column has text
df[df['Notes'].notna()].index

RangeIndex(start=7038, stop=7113, step=1)

In [14]:
# assign first index where 'Notes' has text to `cutoff`
cutoff = df[df['Notes'].notna()].index[0]
cutoff

7038

### Drop metadata

In [15]:
# return df_working excluding rows where 'Notes' column has text
# check last 5 rows
df_working = df.iloc[:cutoff]
df_working.tail()

,Notes,Year,Year Code,State,State Code,Five-Year Age Groups,Five-Year Age Groups Code,Multiple Cause of death,Multiple Cause of death Code,Deaths,Population,Crude Rate
7033,NaN,1999.0,1999.0,Wyoming,56.0,Not Stated,NS,Heroin,T40.1,0,Not Applicable,Not Applicable
7034,NaN,1999.0,1999.0,Wyoming,56.0,Not Stated,NS,Other opioids,T40.2,0,Not Applicable,Not Applicable
7035,NaN,1999.0,1999.0,Wyoming,56.0,Not Stated,NS,Methadone,T40.3,0,Not Applicable,Not Applicable
7036,NaN,1999.0,1999.0,Wyoming,56.0,Not Stated,NS,Other synthetic narcotics,T40.4,0,Not Applicable,Not Applicable
7037,NaN,1999.0,1999.0,Wyoming,56.0,Not Stated,NS,Other and unspecified narcotics,T40.6,0,Not Applicable,Not Applicable


**ADD TO REUSABLE FUNCTION**

In [16]:
# check df_working
df_working.info()

<class 'pandas.DataFrame'>
RangeIndex: 7038 entries, 0 to 7037
Data columns (total 12 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   Notes                         0 non-null      str    
 1   Year                          7038 non-null   float64
 2   Year Code                     7038 non-null   float64
 3   State                         7038 non-null   str    
 4   State Code                    7038 non-null   float64
 5   Five-Year Age Groups          7038 non-null   str    
 6   Five-Year Age Groups Code     7038 non-null   str    
 7   Multiple Cause of death       7038 non-null   str    
 8   Multiple Cause of death Code  7038 non-null   str    
 9   Deaths                        7038 non-null   str    
 10  Population                    7038 non-null   str    
 11  Crude Rate                    7038 non-null   str    
dtypes: float64(3), str(9)
memory usage: 1.1 MB


### Check missing values

In [17]:
df_working.isna().sum()

Notes                           7038
Year                               0
Year Code                          0
State                              0
State Code                         0
Five-Year Age Groups               0
Five-Year Age Groups Code          0
Multiple Cause of death            0
Multiple Cause of death Code       0
Deaths                             0
Population                         0
Crude Rate                         0
dtype: int64

### Check duplicates

In [18]:
df_working.duplicated().sum()

np.int64(0)

### Drop unnecessary columns

In [19]:
# drop 'Notes' column from df_working
df_working = df_working.drop(columns = 'Notes')

**ADD TO REUSABLE FUNCTION**

In [20]:
# check df_working
df_working.info()

<class 'pandas.DataFrame'>
RangeIndex: 7038 entries, 0 to 7037
Data columns (total 11 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   Year                          7038 non-null   float64
 1   Year Code                     7038 non-null   float64
 2   State                         7038 non-null   str    
 3   State Code                    7038 non-null   float64
 4   Five-Year Age Groups          7038 non-null   str    
 5   Five-Year Age Groups Code     7038 non-null   str    
 6   Multiple Cause of death       7038 non-null   str    
 7   Multiple Cause of death Code  7038 non-null   str    
 8   Deaths                        7038 non-null   str    
 9   Population                    7038 non-null   str    
 10  Crude Rate                    7038 non-null   str    
dtypes: float64(3), str(8)
memory usage: 1.0 MB


### Format 'Multiple Cause of death' column name

In [21]:
# capitalize 'Multiple Cause of death' to 'Multiple Cause of Death'
df_working.rename(columns = {
    'Multiple Cause of death':'Multiple Cause of Death',
    'Multiple Cause of death Code':'Multiple Cause of Death Code'
}, inplace = True)

In [22]:
# check df_working
df_working.info()

<class 'pandas.DataFrame'>
RangeIndex: 7038 entries, 0 to 7037
Data columns (total 11 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   Year                          7038 non-null   float64
 1   Year Code                     7038 non-null   float64
 2   State                         7038 non-null   str    
 3   State Code                    7038 non-null   float64
 4   Five-Year Age Groups          7038 non-null   str    
 5   Five-Year Age Groups Code     7038 non-null   str    
 6   Multiple Cause of Death       7038 non-null   str    
 7   Multiple Cause of Death Code  7038 non-null   str    
 8   Deaths                        7038 non-null   str    
 9   Population                    7038 non-null   str    
 10  Crude Rate                    7038 non-null   str    
dtypes: float64(3), str(8)
memory usage: 1.0 MB


## 4. Create reusable function to remove metadata and format

In [23]:
def remove_metadata(filepath: str) -> pd.DataFrame:
    
    # load file
    df = pd.read_csv(filepath, dtype=str)
    
    # assign index of first row with metadata to cutoff
    cutoff = df[df['Notes'].notna()].index[0]
    
    # drop rows with metadata
    df_sans_metadata = df.iloc[:cutoff]
    
    # drop 'Notes' column
    df_sans_metadata = df_sans_metadata.drop(columns='Notes')

    # capitalize 'Multiple Cause of death' to 'Multiple Cause of Death'
    df_sans_metadata.rename(columns = {
        'Multiple Cause of death':'Multiple Cause of Death',
        'Multiple Cause of death Code':'Multiple Cause of Death Code'
    }, inplace = True)

    return df_sans_metadata

# load and concatenate all files
files = [f'../data/raw/YEAR_STATE_AGE_MCD/YEAR_STATE_AGE_MCD_{year}.csv' for year in range(1999, 2021)]
df_clean = pd.concat([remove_metadata(f) for f in files], ignore_index=True)

print(f"Shape: {df_clean.shape}")
print(f"Unique ICD codes: {df_clean['Multiple Cause of Death Code'].unique()}")
print(f"Year range: {df_clean['Year'].min()} – {df_clean['Year'].max()}")

# write to processed directory
df_clean.to_csv(
    '../data/processed/YEAR_STATE_AGE_MCD_1999_2020_preprocessed.csv', 
    index = False
)

print('Saved: ../data/processed/YEAR_STATE_AGE_MCD_1999_2020_preprocessed.csv')

Shape: (154836, 11)
Unique ICD codes: <ArrowStringArray>
['T40.0', 'T40.1', 'T40.2', 'T40.3', 'T40.4', 'T40.6']
Length: 6, dtype: str
Year range: 1999 – 2020
Saved: ../data/processed/YEAR_STATE_AGE_MCD_1999_2020_preprocessed.csv


### Confirm file

In [24]:
os.path.exists('../data/processed/YEAR_STATE_AGE_MCD_1999_2020_preprocessed.csv')

True

## 5. Inspect and validate

In [25]:
# check df_clean
df_clean.info()

<class 'pandas.DataFrame'>
RangeIndex: 154836 entries, 0 to 154835
Data columns (total 11 columns):
 #   Column                        Non-Null Count   Dtype
---  ------                        --------------   -----
 0   Year                          154836 non-null  str  
 1   Year Code                     154836 non-null  str  
 2   State                         154836 non-null  str  
 3   State Code                    154836 non-null  str  
 4   Five-Year Age Groups          154836 non-null  str  
 5   Five-Year Age Groups Code     154836 non-null  str  
 6   Multiple Cause of Death       154836 non-null  str  
 7   Multiple Cause of Death Code  154836 non-null  str  
 8   Deaths                        154836 non-null  str  
 9   Population                    154836 non-null  str  
 10  Crude Rate                    154836 non-null  str  
dtypes: str(11)
memory usage: 24.3 MB


In [26]:
df_clean.head()

,Year,Year Code,State,State Code,Five-Year Age Groups,Five-Year Age Groups Code,Multiple Cause of Death,Multiple Cause of Death Code,Deaths,Population,Crude Rate
0,1999,1999,Alabama,01,< 1 year,1,Opium,T40.0,0,58787,Unreliable
1,1999,1999,Alabama,01,< 1 year,1,Heroin,T40.1,0,58787,Unreliable
2,1999,1999,Alabama,01,< 1 year,1,Other opioids,T40.2,0,58787,Unreliable
3,1999,1999,Alabama,01,< 1 year,1,Methadone,T40.3,0,58787,Unreliable
4,1999,1999,Alabama,01,< 1 year,1,Other synthetic narcotics,T40.4,0,58787,Unreliable


In [27]:
df_clean.tail()

,Year,Year Code,State,State Code,Five-Year Age Groups,Five-Year Age Groups Code,Multiple Cause of Death,Multiple Cause of Death Code,Deaths,Population,Crude Rate
154831,2020,2020,Wyoming,56,Not Stated,NS,Heroin,T40.1,0,Not Applicable,Not Applicable
154832,2020,2020,Wyoming,56,Not Stated,NS,Other opioids,T40.2,0,Not Applicable,Not Applicable
154833,2020,2020,Wyoming,56,Not Stated,NS,Methadone,T40.3,0,Not Applicable,Not Applicable
154834,2020,2020,Wyoming,56,Not Stated,NS,Other synthetic narcotics,T40.4,0,Not Applicable,Not Applicable
154835,2020,2020,Wyoming,56,Not Stated,NS,Other and unspecified narcotics,T40.6,0,Not Applicable,Not Applicable


In [28]:
df_clean.isna().sum()

Year                            0
Year Code                       0
State                           0
State Code                      0
Five-Year Age Groups            0
Five-Year Age Groups Code       0
Multiple Cause of Death         0
Multiple Cause of Death Code    0
Deaths                          0
Population                      0
Crude Rate                      0
dtype: int64

In [29]:
df_clean.duplicated().sum()

np.int64(0)

In [30]:
df_clean.nunique()

Year                               22
Year Code                          22
State                              51
State Code                         51
Five-Year Age Groups               23
Five-Year Age Groups Code          23
Multiple Cause of Death             6
Multiple Cause of Death Code        6
Deaths                            356
Population                      19782
Crude Rate                        630
dtype: int64

## 6. Column validation and cleaning

### `Year` and `YEAR CODE`

In [31]:
# check unique values in 'Year'
df_clean['Year'].unique()

<ArrowStringArray>
['1999', '2000', '2001', '2002', '2003', '2004', '2005', '2006', '2007',
 '2008', '2009', '2010', '2011', '2012', '2013', '2014', '2015', '2016',
 '2017', '2018', '2019', '2020']
Length: 22, dtype: str

In [32]:
# check unique values in 'Year Code'
df_clean['Year Code'].unique()

<ArrowStringArray>
['1999', '2000', '2001', '2002', '2003', '2004', '2005', '2006', '2007',
 '2008', '2009', '2010', '2011', '2012', '2013', '2014', '2015', '2016',
 '2017', '2018', '2019', '2020']
Length: 22, dtype: str

In [33]:
# check if 'Year' == 'Year Code'
(df_clean['Year'] == df_clean['Year Code']).all()

np.True_

In [34]:
# convert 'Year' column to numerical values
df_clean['Year'] = df_clean['Year'].astype('int64')

**ADD TO REUSABLE FUNCTION**

In [35]:
# drop 'Year Code' column
df_clean.drop(columns = ['Year Code'], inplace = True)

**ADD TO REUSABLE FUNCTION**

In [36]:
# check df_clean
df_clean.info()

<class 'pandas.DataFrame'>
RangeIndex: 154836 entries, 0 to 154835
Data columns (total 10 columns):
 #   Column                        Non-Null Count   Dtype
---  ------                        --------------   -----
 0   Year                          154836 non-null  int64
 1   State                         154836 non-null  str  
 2   State Code                    154836 non-null  str  
 3   Five-Year Age Groups          154836 non-null  str  
 4   Five-Year Age Groups Code     154836 non-null  str  
 5   Multiple Cause of Death       154836 non-null  str  
 6   Multiple Cause of Death Code  154836 non-null  str  
 7   Deaths                        154836 non-null  str  
 8   Population                    154836 non-null  str  
 9   Crude Rate                    154836 non-null  str  
dtypes: int64(1), str(9)
memory usage: 21.9 MB


### `State` and `State Code` 

In [37]:
# check unique values in 'State'
df_clean['State'].unique()

<ArrowStringArray>
[             'Alabama',               'Alaska',              'Arizona',
             'Arkansas',           'California',             'Colorado',
          'Connecticut',             'Delaware', 'District of Columbia',
              'Florida',              'Georgia',               'Hawaii',
                'Idaho',             'Illinois',              'Indiana',
                 'Iowa',               'Kansas',             'Kentucky',
            'Louisiana',                'Maine',             'Maryland',
        'Massachusetts',             'Michigan',            'Minnesota',
          'Mississippi',             'Missouri',              'Montana',
             'Nebraska',               'Nevada',        'New Hampshire',
           'New Jersey',           'New Mexico',             'New York',
       'North Carolina',         'North Dakota',                 'Ohio',
             'Oklahoma',               'Oregon',         'Pennsylvania',
         'Rhode Island',       '

In [38]:
# check unique values in 'State Code'
df_clean['State Code'].unique()

<ArrowStringArray>
['01', '02', '04', '05', '06', '08', '09', '10', '11', '12', '13', '15', '16',
 '17', '18', '19', '20', '21', '22', '23', '24', '25', '26', '27', '28', '29',
 '30', '31', '32', '33', '34', '35', '36', '37', '38', '39', '40', '41', '42',
 '44', '45', '46', '47', '48', '49', '50', '51', '53', '54', '55', '56']
Length: 51, dtype: str

### `Five-Year Age Groups` and `Five-Year Age Groups Code`

In [39]:
# check unique values in 'Five-Year Age Groups'
df_clean['Five-Year Age Groups'].unique()

<ArrowStringArray>
[    '< 1 year',    '1-4 years',    '5-9 years',  '10-14 years',
  '15-19 years',  '20-24 years',  '25-29 years',  '30-34 years',
  '35-39 years',  '40-44 years',  '45-49 years',  '50-54 years',
  '55-59 years', '60-64 years ',  '65-69 years',  '70-74 years',
  '75-79 years',  '80-84 years',  '85-89 years',  '90-94 years',
  '95-99 years',   '100+ years',   'Not Stated']
Length: 23, dtype: str

In [40]:
# check unique values in 'Five-Year Age Groups Code'
df_clean['Five-Year Age Groups Code'].unique()

<ArrowStringArray>
[    '1',   '1-4',   '5-9', '10-14', '15-19', '20-24', '25-29', '30-34',
 '35-39', '40-44', '45-49', '50-54', '55-59', '60-64', '65-69', '70-74',
 '75-79', '80-84', '85-89', '90-94', '95-99',  '100+',    'NS']
Length: 23, dtype: str

In [41]:
# temporarily strip 'Five-Year Age Groups' to match 'Five-Year Age Groups Code'
(df_clean['Five-Year Age Groups']
    .str.replace('< ', '')
    .str.replace(' years?', '', regex = True)
    .str.replace('Not Stated', 'NS')
    .str.strip()
    .unique()
)

<ArrowStringArray>
[    '1',   '1-4',   '5-9', '10-14', '15-19', '20-24', '25-29', '30-34',
 '35-39', '40-44', '45-49', '50-54', '55-59', '60-64', '65-69', '70-74',
 '75-79', '80-84', '85-89', '90-94', '95-99',  '100+',    'NS']
Length: 23, dtype: str

In [42]:
# check both columns against each other
((df_clean['Five-Year Age Groups']
    .str.replace('< ', '')
    .str.replace(' years?', '', regex = True)
    .str.replace('Not Stated', 'NS')
    .str.strip()
) == df_clean['Five-Year Age Groups Code']).all()

np.True_

In [43]:
# drop 'Five-Year Age Groups' column
df_clean.drop(columns = ['Five-Year Age Groups'], inplace = True)

**ADD TO REUSABLE FUNCTION**

In [44]:
# check unique values in 'Five-Year Age Groups' column
df_clean['Five-Year Age Groups Code'].unique()

<ArrowStringArray>
[    '1',   '1-4',   '5-9', '10-14', '15-19', '20-24', '25-29', '30-34',
 '35-39', '40-44', '45-49', '50-54', '55-59', '60-64', '65-69', '70-74',
 '75-79', '80-84', '85-89', '90-94', '95-99',  '100+',    'NS']
Length: 23, dtype: str

In [45]:
# check unique number of values across all columns
df_clean.nunique()

Year                               22
State                              51
State Code                         51
Five-Year Age Groups Code          23
Multiple Cause of Death             6
Multiple Cause of Death Code        6
Deaths                            356
Population                      19782
Crude Rate                        630
dtype: int64

In [46]:
# check df_clean
df_clean.info()

<class 'pandas.DataFrame'>
RangeIndex: 154836 entries, 0 to 154835
Data columns (total 9 columns):
 #   Column                        Non-Null Count   Dtype
---  ------                        --------------   -----
 0   Year                          154836 non-null  int64
 1   State                         154836 non-null  str  
 2   State Code                    154836 non-null  str  
 3   Five-Year Age Groups Code     154836 non-null  str  
 4   Multiple Cause of Death       154836 non-null  str  
 5   Multiple Cause of Death Code  154836 non-null  str  
 6   Deaths                        154836 non-null  str  
 7   Population                    154836 non-null  str  
 8   Crude Rate                    154836 non-null  str  
dtypes: int64(1), str(8)
memory usage: 19.1 MB


### `Multiple Cause of Death` and `Multiple Cause of Death Code`

In [47]:
# check unique values in 'Multiple Cause of Death' column
df_clean['Multiple Cause of Death'].unique()

<ArrowStringArray>
[                          'Opium',                          'Heroin',
                   'Other opioids',                       'Methadone',
       'Other synthetic narcotics', 'Other and unspecified narcotics']
Length: 6, dtype: str

In [48]:
# check unique values in 'Multiple Cause of Death Code' column
df_clean['Multiple Cause of Death Code'].unique()

<ArrowStringArray>
['T40.0', 'T40.1', 'T40.2', 'T40.3', 'T40.4', 'T40.6']
Length: 6, dtype: str

### `Deaths`

In [49]:
# check total number of unique values in 'Deaths' column
df_clean['Deaths'].nunique()

356

In [50]:
# check first five rows in 'Deaths' column
df_clean['Deaths'].head()

0    0
1    0
2    0
3    0
4    0
Name: Deaths, dtype: str

In [51]:
# check last five rows in 'Deaths' column
df_clean['Deaths'].tail()

154831    0
154832    0
154833    0
154834    0
154835    0
Name: Deaths, dtype: str

In [52]:
# convert values in 'Deaths' column into numerical values
df_clean['Deaths'] = df_clean['Deaths'].astype('int64')

ValueError: invalid literal for int() with base 10: 'Suppressed'

**UNEXPECTED ERROR - SOME VALUES CANNOT BE CONVERTED INTO NUMERICAL VALUE**

In [ ]:
# convert values in 'Deaths' column to numerical values, otherwise converts to NaN
pd.to_numeric(df_clean['Deaths'], errors = 'coerce')

In [53]:
# returns values from 'Deaths' column where values = NaN (False)
pd.to_numeric(df_clean['Deaths'], errors = 'coerce').isna()

0         False
1         False
2         False
3         False
4         False
          ...  
154831    False
154832    False
154833    False
154834    False
154835    False
Name: Deaths, Length: 154836, dtype: bool

In [54]:
# return rows where 'Deaths' = NaN
df_clean[pd.to_numeric(df_clean['Deaths'], errors = 'coerce').isna()].head()

,Year,State,State Code,Five-Year Age Groups Code,Multiple Cause of Death,Multiple Cause of Death Code,Deaths,Population,Crude Rate
14,1999,Alabama,01,5-9,Other opioids,T40.2,Suppressed,315749,Suppressed
37,1999,Alabama,01,25-29,Heroin,T40.1,Suppressed,307028,Suppressed
38,1999,Alabama,01,25-29,Other opioids,T40.2,Suppressed,307028,Suppressed
39,1999,Alabama,01,25-29,Methadone,T40.3,Suppressed,307028,Suppressed
40,1999,Alabama,01,25-29,Other synthetic narcotics,T40.4,Suppressed,307028,Suppressed


**THERE ARE OBJECTS IN THE `Deaths` COLUMN WHICH PREVENT NUMERICAL CONVERSION**

In [55]:
# return unique values in 'Deaths' column where NaN is False
(df_clean[pd.to_numeric(df_clean['Deaths'], errors = 'coerce')
    .isna()]['Deaths']
    .unique()
)

<ArrowStringArray>
['Suppressed']
Length: 1, dtype: str

**`Suppressed` PREVENTS NUMERICAL CONVERSION**

In [56]:
# return sum of 'Suppressed' occurrences in 'Deaths' column
(df_clean['Deaths'] == 'Suppressed').sum()

np.int64(36100)

**TOTAL OF 36100 INSTANCES WHERE 'Deaths' is 'Suppressed'**

In [57]:
# add 'Deaths_Supressed' column before replacing 'Suppressed' with NaN
df_clean['Deaths_Suppressed'] = df_clean['Deaths'] == 'Suppressed'

**ADD TO REUSABLE FUNCTION**

In [58]:
# check unique values in `Deaths_Suppressed` column
df_clean['Deaths_Suppressed'].unique()

array([False,  True])

In [59]:
df_clean.info()

<class 'pandas.DataFrame'>
RangeIndex: 154836 entries, 0 to 154835
Data columns (total 10 columns):
 #   Column                        Non-Null Count   Dtype
---  ------                        --------------   -----
 0   Year                          154836 non-null  int64
 1   State                         154836 non-null  str  
 2   State Code                    154836 non-null  str  
 3   Five-Year Age Groups Code     154836 non-null  str  
 4   Multiple Cause of Death       154836 non-null  str  
 5   Multiple Cause of Death Code  154836 non-null  str  
 6   Deaths                        154836 non-null  str  
 7   Population                    154836 non-null  str  
 8   Crude Rate                    154836 non-null  str  
 9   Deaths_Suppressed             154836 non-null  bool 
dtypes: bool(1), int64(1), str(8)
memory usage: 19.2 MB


In [60]:
# return values in 'Deaths' colummn where 'Deaths' is 'Suppressed'
df_clean.loc[df_clean['Deaths'] == 'Suppressed', 'Deaths']

14        Suppressed
37        Suppressed
38        Suppressed
39        Suppressed
40        Suppressed
             ...    
154783    Suppressed
154784    Suppressed
154786    Suppressed
154787    Suppressed
154796    Suppressed
Name: Deaths, Length: 36100, dtype: str

In [61]:
# assign NaN to 'Suppressed' values in 'Deaths' column 
df_clean.loc[df_clean['Deaths'] == 'Suppressed', 'Deaths'] = np.nan

**ADD TO REUSABLE FUNCITON**

In [62]:
# return sum of NaN occurences in 'Deaths' column
df_clean['Deaths'].isna().sum()

np.int64(36100)

In [63]:
# check df_clean
df_clean.info()

<class 'pandas.DataFrame'>
RangeIndex: 154836 entries, 0 to 154835
Data columns (total 10 columns):
 #   Column                        Non-Null Count   Dtype
---  ------                        --------------   -----
 0   Year                          154836 non-null  int64
 1   State                         154836 non-null  str  
 2   State Code                    154836 non-null  str  
 3   Five-Year Age Groups Code     154836 non-null  str  
 4   Multiple Cause of Death       154836 non-null  str  
 5   Multiple Cause of Death Code  154836 non-null  str  
 6   Deaths                        118736 non-null  str  
 7   Population                    154836 non-null  str  
 8   Crude Rate                    154836 non-null  str  
 9   Deaths_Suppressed             154836 non-null  bool 
dtypes: bool(1), int64(1), str(8)
memory usage: 18.9 MB


In [64]:
# convert values in 'Deaths' column into numerical values
df_clean['Deaths'] = pd.to_numeric(df_clean['Deaths'], errors='coerce')
df_clean['Deaths'].dtype

dtype('float64')

**ADD TO REUSABLE FUNCTION**

In [65]:
# check df_clean
df_clean.info()

<class 'pandas.DataFrame'>
RangeIndex: 154836 entries, 0 to 154835
Data columns (total 10 columns):
 #   Column                        Non-Null Count   Dtype  
---  ------                        --------------   -----  
 0   Year                          154836 non-null  int64  
 1   State                         154836 non-null  str    
 2   State Code                    154836 non-null  str    
 3   Five-Year Age Groups Code     154836 non-null  str    
 4   Multiple Cause of Death       154836 non-null  str    
 5   Multiple Cause of Death Code  154836 non-null  str    
 6   Deaths                        118736 non-null  float64
 7   Population                    154836 non-null  str    
 8   Crude Rate                    154836 non-null  str    
 9   Deaths_Suppressed             154836 non-null  bool   
dtypes: bool(1), float64(1), int64(1), str(7)
memory usage: 18.7 MB


### `Population`

In [66]:
# convert values in 'Population' column into numerical values
df_clean['Population'] = df_clean['Population'].astype('int64')

ValueError: invalid literal for int() with base 10: 'Not Applicable'

**UNEXPECTED ERROR - SOME VALUES CANNOT BE CONVERTED INTO NUMERICAL VALUE**

In [67]:
# return unique values in 'Population' column where NaN is False
(df_clean[pd.to_numeric(df_clean['Population'], errors = 'coerce')
    .isna()]['Population']
    .unique()
)

<ArrowStringArray>
['Not Applicable']
Length: 1, dtype: str

**`Not Applicable` PREVENTS NUMERICAL CONVERSION**

In [68]:
# return sum of 'Not Applicable' occurrences in 'Population' column
(df_clean['Population'] == 'Not Applicable').sum()

np.int64(33660)

**TOTAL OF 33660 INSTANCES WHERE 'Population' is 'Not Applicable'**

In [69]:
# check if 'Deaths_Suppressed' overlaps with 'Not Applicable' in the 'Population' column
df_clean[df_clean['Population'] == 'Not Applicable']['Deaths_Suppressed'].value_counts()

Deaths_Suppressed
False    32513
True      1147
Name: count, dtype: int64

**Not Applicable' DOES NOT OVERLAP WITH 'Deaths_Suppressed'**

In [70]:
# check if unique values of 'Five-Year Age Groups Code' overlaps with 
# 'Not Applicable' in the 'Population' column
df_clean[df_clean['Population'] == 'Not Applicable']['Five-Year Age Groups Code'].unique()

<ArrowStringArray>
['85-89', '90-94', '95-99', '100+', 'NS']
Length: 5, dtype: str

**'Not Applicable' DOES OVERLAP WITH THE FOLLOWING 'Five-Year Age Groups'**

In [71]:
# count total rows from the 'Population' column
# where 'Five-Year Age Groups Code' meets the following age groups
(df_clean[df_clean['Five-Year Age Groups Code']
    .isin(['85-89', '90-94', '95-99', '100+', 'NS'])]['Population']
    .value_counts()
)

Population
Not Applicable    33660
Name: count, dtype: int64

**TOTAL OF 33660 INSTANCES WHERE 'Five-Year Age Groups Code' OVERLAP WITH 'Not Applicable' from 'Population'**

In [72]:
# check count of 'Deaths' where 'Population' is 'Not Applicable'
df_clean[df_clean['Population'] == 'Not Applicable']['Deaths'].value_counts()

Deaths
0.0     32512
10.0        1
Name: count, dtype: int64

**ONE VALUE OF 'Deaths' WHERE 'Population' IS 'Not Applicable'**

In [73]:
# check row(s) where 'Population' is 'Not Applcable' and 'Deaths' is 10
df_clean[
    (df_clean['Population'] == 'Not Applicable') & 
    (df_clean['Deaths'] == 10.0)
]

,Year,State,State Code,Five-Year Age Groups Code,Multiple Cause of Death,Multiple Cause of Death Code,Deaths,Population,Crude Rate,Deaths_Suppressed
148460,2020,California,06,85-89,Other opioids,T40.2,10.0,Not Applicable,Not Applicable,False


In [74]:
df_clean[df_clean['Population'] == 'Not Applicable']

,Year,State,State Code,Five-Year Age Groups Code,Multiple Cause of Death,Multiple Cause of Death Code,Deaths,Population,Crude Rate,Deaths_Suppressed
108,1999,Alabama,01,85-89,Opium,T40.0,0.0,Not Applicable,Not Applicable,False
109,1999,Alabama,01,85-89,Heroin,T40.1,0.0,Not Applicable,Not Applicable,False
110,1999,Alabama,01,85-89,Other opioids,T40.2,0.0,Not Applicable,Not Applicable,False
111,1999,Alabama,01,85-89,Methadone,T40.3,0.0,Not Applicable,Not Applicable,False
112,1999,Alabama,01,85-89,Other synthetic narcotics,T40.4,0.0,Not Applicable,Not Applicable,False
...,...,...,...,...,...,...,...,...,...,...
154831,2020,Wyoming,56,NS,Heroin,T40.1,0.0,Not Applicable,Not Applicable,False
154832,2020,Wyoming,56,NS,Other opioids,T40.2,0.0,Not Applicable,Not Applicable,False
154833,2020,Wyoming,56,NS,Methadone,T40.3,0.0,Not Applicable,Not Applicable,False
154834,2020,Wyoming,56,NS,Other synthetic narcotics,T40.4,0.0,Not Applicable,Not Applicable,False


In [75]:
# keep only rows where 'Population' is not 'Not Applicable'
df_clean = df_clean[df_clean['Population'] != 'Not Applicable']

**ADD TO REUSABLE FUNCTION**

In [76]:
# check df_clean
df_clean.info()

<class 'pandas.DataFrame'>
Index: 121176 entries, 0 to 154805
Data columns (total 10 columns):
 #   Column                        Non-Null Count   Dtype  
---  ------                        --------------   -----  
 0   Year                          121176 non-null  int64  
 1   State                         121176 non-null  str    
 2   State Code                    121176 non-null  str    
 3   Five-Year Age Groups Code     121176 non-null  str    
 4   Multiple Cause of Death       121176 non-null  str    
 5   Multiple Cause of Death Code  121176 non-null  str    
 6   Deaths                        86223 non-null   float64
 7   Population                    121176 non-null  str    
 8   Crude Rate                    121176 non-null  str    
 9   Deaths_Suppressed             121176 non-null  bool   
dtypes: bool(1), float64(1), int64(1), str(7)
memory usage: 15.3 MB


In [77]:
# convert values in 'Population' column into numerical values
df_clean['Population'] = pd.to_numeric(df_clean['Population'], errors='coerce')
df_clean['Population'].dtype

dtype('int64')

**ADD TO REUSABLE FUNCTION**

In [78]:
# check df_clean
df_clean.info()

<class 'pandas.DataFrame'>
Index: 121176 entries, 0 to 154805
Data columns (total 10 columns):
 #   Column                        Non-Null Count   Dtype  
---  ------                        --------------   -----  
 0   Year                          121176 non-null  int64  
 1   State                         121176 non-null  str    
 2   State Code                    121176 non-null  str    
 3   Five-Year Age Groups Code     121176 non-null  str    
 4   Multiple Cause of Death       121176 non-null  str    
 5   Multiple Cause of Death Code  121176 non-null  str    
 6   Deaths                        86223 non-null   float64
 7   Population                    121176 non-null  int64  
 8   Crude Rate                    121176 non-null  str    
 9   Deaths_Suppressed             121176 non-null  bool   
dtypes: bool(1), float64(1), int64(2), str(6)
memory usage: 14.6 MB


### `Crude Rate`

In [79]:
# check unique values for 'Crude Rate'
df_clean['Crude Rate'].nunique()

629

In [80]:
# convert values in 'Crude Rate' column into numerical values
df_clean['Crude Rate'] = df_clean['Crude Rate'].astype('float64')

ValueError: could not convert string to float: 'Unreliable'

**UNEXPECTED ERROR - SOME VALUES CANNOT BE CONVERTED INTO NUMERICAL VALUE**

In [81]:
# return unique values in 'Crude Rate' column where NaN is False
(df_clean[pd.to_numeric(df_clean['Crude Rate'], errors='coerce')
    .isna()]['Crude Rate']
    .unique()
)

<ArrowStringArray>
['Unreliable', 'Suppressed']
Length: 2, dtype: str

**`Unreliable` AND `Suppressed` PREVENTS NUMERICAL CONVERSION**

In [82]:
# return sum counts of 'Suppressed' + 'Unreliable'
df_clean['Crude Rate'].isin(['Suppressed', 'Unreliable']).sum()

np.int64(112184)

In [83]:
# return counts where 'Crude Rate' is either 'Unreliable' or 'Suppressed'
(df_clean[pd.to_numeric(df_clean['Crude Rate'], errors='coerce')
    .isna()]['Crude Rate']
    .value_counts()
)

Crude Rate
Unreliable    77231
Suppressed    34953
Name: count, dtype: int64

**TOTAL OF 112184 INSTANCES WHERE 'Crude Rate' is 'Unreliable' or 'Suppressed**

In [84]:
# calculate percentage of 'Unreliable' + 'Suppressed' from 121176 total rows
non_numeric = pd.to_numeric(df_clean['Crude Rate'], errors='coerce').isna().sum()

total = len(df_clean)

print(f"{non_numeric / total * 100:.2f}% of 'Crude Rate' is either 'Unreliable' or 'Suppressed'")

92.58% of 'Crude Rate' is either 'Unreliable' or 'Suppressed'


In [85]:
# create 'Crude Rate Calculated' column
df_clean['Crude Rate Calculated'] = (df_clean['Deaths'] / df_clean['Population']) * 100000

**ADD TO REUSUABLE FUNCTION**

In [86]:
# drop 'Crude Rate' column
df_clean.drop(columns = ['Crude Rate'], inplace = True)

**ADD TO REUSUABLE FUNCTION**

In [87]:
# check df_clean
df_clean.info()

<class 'pandas.DataFrame'>
Index: 121176 entries, 0 to 154805
Data columns (total 10 columns):
 #   Column                        Non-Null Count   Dtype  
---  ------                        --------------   -----  
 0   Year                          121176 non-null  int64  
 1   State                         121176 non-null  str    
 2   State Code                    121176 non-null  str    
 3   Five-Year Age Groups Code     121176 non-null  str    
 4   Multiple Cause of Death       121176 non-null  str    
 5   Multiple Cause of Death Code  121176 non-null  str    
 6   Deaths                        86223 non-null   float64
 7   Population                    121176 non-null  int64  
 8   Deaths_Suppressed             121176 non-null  bool   
 9   Crude Rate Calculated         86223 non-null   float64
dtypes: bool(1), float64(2), int64(2), str(5)
memory usage: 13.5 MB


In [88]:
# reset index
df_clean = df_clean.reset_index(drop=True)

**ADD TO REUSUABLE FUNCTION**

## 7. Export cleaned data

In [89]:
print(f"Shape: {df_clean.shape}")
print(f"Unique ICD codes: {df_clean['Multiple Cause of Death Code'].unique()}")
print(f"Year range: {df_clean['Year'].min()} – {df_clean['Year'].max()}")

# write to processed directory
df_clean.to_csv(
    '../data/processed/YEAR_STATE_AGE_MCD_1999_2020_cleaned.csv', 
    index = False
)

print('Saved: ../data/processed/MCD_YEAR_STATE_AGE_MCD_1999_2020_cleaned.csv')

Shape: (121176, 10)
Unique ICD codes: <ArrowStringArray>
['T40.0', 'T40.1', 'T40.2', 'T40.3', 'T40.4', 'T40.6']
Length: 6, dtype: str
Year range: 1999 – 2020
Saved: ../data/processed/MCD_YEAR_STATE_AGE_MCD_1999_2020_cleaned.csv


In [90]:
os.path.exists('../data/processed/YEAR_STATE_AGE_MCD_1999_2020_cleaned.csv')

True